working/terminating-tokens/feature-specification.md

```diff
@@ -0,0 +1,1656 @@
+    the `;` key.
+
+*   **As an output format for code generators.** The Dart ecosystem uses code
+    generation pretty heavily. It's often easier for tools that output Dart
+    source code to be able to do so using semicolons without worrying about
+    whitespace or formatting. Those tools can then either run the output through
+    dartfmt to remove the semicolons or just leave them since they're harmless.
+
+OK, that's what I have to try to persuade you that optional semicolon are an
+idea worth *wanting*. The next question is whether it's an *attainable* goal.
+JavaScript is a good example of the fact that optional semicolons can be done,
+but done so poorly as to be an anti-feature.
+
+Also, each language has its own pecularities, so a good optional semicolon
+system in one language may not work in another. In other words, we need to not
+just specify the optional semicolon rules, but *evaluate* them in the context
+Dart's own grammar and ecosystem.
+
+## Evaluation
+
+Given some set of rules for eliminating semicolons, how can we tell if they are
+*good* rules? I measure it along a couple of axes. We want to maximize:
+
+*   **Unambiguity** - Removing mandatory semicolons as terminators can lead to
+    ambiguous cases where both ignoring a newline or treating it as a semicolon
+    lead to valid parses. Any syntax *must* be unambiguous&mdash;users don't
+    like it if their code is parsed differently on different days of the week.
+    When given an ambiguous choice, it must resolve it in a way that aligns with
+    the user's intention as often as possible.
+
+*   **Compatibility** - How much existing code (with explicit semicolons and
+    whatever whitespace and newlines it happens to have) parses the same way
+    with the new newline-sensitive grammar as it did with the old grammar? Any
+    difference here is existing code that is broken by the new rules.
+
+*   **Accuracy** - When users write code without semicolons and with newlines
+    wherever they feel they look best, how often does the parser see the code
+    they way the user wants it to? How often do users have to fight the grammar
+    and tweak newlines to satisfy the parser?
+
+*   **Robustness** - Given a chunk of code where the newlines are *not*
+    deliberate, how likely is it to be parsed the way the user wants? If someone
+    is just banging code into their editor and not thinking about
+    formatting&mdash;that's dartfmt's job after all&mdash;how likely are they to
+    get what they want out? If they copy and paste a big chunk of code somewhere else, how often does it end up still correct?
+
+In an ideal world, we'd hit 100% on all four of those. Short of radically
+changing the language into something like Lisp, that's impossible. But we can
+run experiments on a corpus of Dart code to try to empirically measure these to
+some degree.
+
+The corpus I used is the Flutter repository, the Dart SDK repository, and the
```

Comment:
Why mention Flutter? Yes, it uses Dart, but so does GoogleAdware and Google Adsense. Additionally, there's the AngularDart team, plus all the frameworks the community has built.

Is this change being requested solely to benefit Flutter uses? I would urge that either these other uses of Dart are included or otherwise limit the mentions of Flutter so it's not so specific to one user base. Though you do go through packages on Pub, so I don't know why Flutter had to specifically be called out and the rest excluded.

Reply:
This change is part of the "UI as code" charter, which is detailed more [here](https://github.com/munificent/ui-as-code) and [here](https://github.com/munificent/ui-as-code/blob/master/Motivation.md). It's not *only* to benefit Flutter, but if it *doesn't* benefit Flutter, then I don't think we have the resources to work on it right now. In other words "useful for Flutter" is a necessary condition, but we certainly hope that all "UI as code" features are useful for much *more* than Flutter too.

Flutter's repo is an important use case because it's a very large, very popular corpus of Dart code and is also one that does not currently use dartfmt and has its own formatting style. So it's a good test case for how robust the optional semicolon rules are across a variety of Dart code.

Reply:
Hmm, thanks for the quick response @munificent. I appreciate the openness and you provided some clarity